### PR TITLE
application offline payments

### PR DESCRIPTION
**Pull request by vinil-transerve**

### CODE REVIEW

File: estate-services/src/main/java/org/egov/ps/config/Configuration.java [hunk 1]
```diff
@@ -137,4 +137,16 @@ public MappingJackson2HttpMessageConverter jacksonConverter(ObjectMapper objectM
...
+	 * Collection Service
+	 */
+	@Value("${egov.collectionservice.host}")
+	private String collectionPaymentHost;
+
+	@Value("${egov.collectionservice.payment.create.path}")
+	private String collectionPaymentEndPoint;
+	
+	@Value("${egov.bill.gen.endpoint}")
+	private String billGenearateEndpoint;
```
Comment (rajiv-transerve) [inline estate-services/src/main/java/org/egov/ps/config/Configuration.java]: Change to `billGenerateEndpoint`

File: estate-services/src/main/java/org/egov/ps/service/ApplicationService.java [hunk 2]
```diff
@@ -97,4 +106,39 @@
...
+		 * create an offline payment.
+		 */
+		demandService.createCashPayment(applicationRequest.getRequestInfo(), application.getPaymentAmount(),
+				bills.get(0).getId(), application, application.getBillingBusinessService());
+
+		OfflinePaymentDetails offlinePaymentDetails = OfflinePaymentDetails.builder().id(UUID.randomUUID().toString())
+				.propertyId(application.getId()).demandId(bills.get(0).getBillDetails().get(0).getDemandId())
+				.amount(application.getPaymentAmount()).bankName(application.getBankName())
+				.transactionNumber(application.getTransactionId()).build();
+		application.setOfflinePaymentDetails(Collections.singletonList(offlinePaymentDetails));
```
Comment (rajiv-transerve) [inline estate-services/src/main/java/org/egov/ps/service/ApplicationService.java]: To be removed

File: estate-services/src/main/java/org/egov/ps/validator/ApplicationValidatorService.java [hunk 3]
```diff
@@ -243,7 +243,7 @@ private boolean ApplicationExists(List<Application> applicationRequest) {
 		return (!CollectionUtils.isEmpty(applicationRequest) && applicationRequest.size() == 1);
 	}
 
-	private ApplicationCriteria getApplicationCriteria(ApplicationRequest request) {
+	public ApplicationCriteria getApplicationCriteria(ApplicationRequest request) {
```
Comment (rajiv-transerve) [inline estate-services/src/main/java/org/egov/ps/validator/ApplicationValidatorService.java]: revert

**Review by rajiv-transerve — CHANGES_REQUESTED** [review]

File: estate-services/src/main/java/org/egov/ps/controller/ApplicationController.java [hunk 4]
```diff
@@ -84,4 +88,16 @@
 		return new ResponseEntity<>(response, HttpStatus.OK);
 
 	}
+
+	@PostMapping("/_offlinepayment")
```
Comment (rajiv-transerve) [inline estate-services/src/main/java/org/egov/ps/controller/ApplicationController.java]: Change this to `_collect_payment` and make sure only employees can access these requests in MDMS.

File: estate-services/src/main/java/org/egov/ps/model/Application.java [hunk 5]
```diff
@@ -202,4 +205,32 @@ public Application addApplicationDocumentsItem(Document applicationDocumentItem)
...
+	@Size(max = 256, message = "bank name must be between 0 and 256 characters in length")
+	private String bankName;
+	
+	@JsonProperty("transactionId")
+	@Size(max = 256, message = "transaction id must be between 0 and 256 characters in length")
+	private String transactionId;
+	
+	@Valid
+	@JsonProperty
+	private List<OfflinePaymentDetails> offlinePaymentDetails;
```
Comment (rajiv-transerve) [inline estate-services/src/main/java/org/egov/ps/model/Application.java]: This should not be there. As mentioned yesterday, there will only ever be a single payment.

File: estate-services/src/main/java/org/egov/ps/model/OfflinePaymentDetails.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
+package org.egov.ps.model;
```
Comment (rajiv-transerve) [inline estate-services/src/main/java/org/egov/ps/model/OfflinePaymentDetails.java]: This file can be retained but only to be used for collecting offline rent for properties not application offline payment.

File: estate-services/src/main/java/org/egov/ps/model/Application.java [hunk 7]
```diff
@@ -202,4 +205,32 @@ public Application addApplicationDocumentsItem(Document applicationDocumentItem)
...
 		return String.format("%s_%s_%s", this.getBranchType(), this.getModuleType(), this.getApplicationType());
 	}
+	
+	/**
+	 * Pending consumer code. This needs to be saved in the database for online
+	 * payments.
+	 */
+	@JsonProperty("paymentConsumerCode")
+	@Size(max = 256, message = "payment consumer code must be between 0 and 256 characters in length")
+	private String paymentConsumerCode;
```
Comment (rajiv-transerve) [inline estate-services/src/main/java/org/egov/ps/model/Application.java]: There should be no new paymentConsumerCode for an application. The application number should be used to generated demand, bill etc..

**Review by rajiv-transerve — CHANGES_REQUESTED** [review]

File: estate-services/src/main/java/org/egov/ps/model/Application.java (hunk 5 above)
Comment (vinil-transerve) [reply estate-services/src/main/java/org/egov/ps/model/Application.java]: yes you said, sorry, I will do it know

File: estate-services/src/main/java/org/egov/ps/config/Configuration.java (hunk 1 above)
Comment (vinil-transerve) [reply estate-services/src/main/java/org/egov/ps/config/Configuration.java]: done

File: estate-services/src/main/java/org/egov/ps/service/ApplicationService.java (hunk 2 above)
Comment (vinil-transerve) [reply estate-services/src/main/java/org/egov/ps/service/ApplicationService.java]: done

File: estate-services/src/main/java/org/egov/ps/validator/ApplicationValidatorService.java (hunk 3 above)
Comment (vinil-transerve) [reply estate-services/src/main/java/org/egov/ps/validator/ApplicationValidatorService.java]: done

File: estate-services/src/main/java/org/egov/ps/controller/ApplicationController.java (hunk 4 above)
Comment (vinil-transerve) [reply estate-services/src/main/java/org/egov/ps/controller/ApplicationController.java]: yes, has to add in mdms

**Pull request merged**